source/azure_iot.c

```diff
@@ -0,0 +1,455 @@
+            ulStep++;
+        }
+        else if( ulStep == 1 )
+        {
+            /* Use last 2 bits of encoded bytes character and first 4 bits of next encoded bytes character for index.  */
+            pcEncodedBytes[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( ( uint8_t ) pucBytes[ i ] ) & 0x3 ) << 4 ) |
+                                                                    ( ( ( uint8_t ) pucBytes[ i + 1 ] ) >> 4 ) ];
+            i++;
+            ulStep++;
+        }
+        else if( ulStep == 2 )
+        {
+            /* Use last 4 bits of encoded bytes character and first 2 bits of next encoded bytes character for index.  */
+            pcEncodedBytes[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( ( uint8_t ) pucBytes[ i ] ) & 0xF ) << 2 ) |
+                                                                    ( ( ( uint8_t ) pucBytes[ i + 1 ] ) >> 6 ) ];
+            i++;
+            ulStep++;
+        }
+        else /* Step 3 */
+        {
+            /* Use last 6 bits of encoded bytes character for index.  */
+            pcEncodedBytes[ j++ ] = ( char ) _cAzureIoTBase64Array[ ( ( ( uint8_t ) pucBytes[ i ] ) & 0x3F ) ];
+            i++;
+            ulStep = 0;
+        }
+    }
+
+    /* Determine if the index needs to be advanced.  */
+    if( ulStep != 3 )
+    {
+        i++;
+    }
+
+    /* Now add the PAD characters.  */
+    while( ( ulPad-- ) && ( j < ulEncodedBytesLength ) )
+    {
+        /* Pad pcEncodedBytes with '=' characters.  */
+        pcEncodedBytes[ j++ ] = '=';
+    }
+
+    /* Need space for a NULL terminator */
+    if( j >= ulEncodedBytesLength )
+    {
+        return eAzureIoTOutOfMemory;
+    }
+
+    /* Put a NULL character in.  */
+    pcEncodedBytes[ j ] = 0;
+    *pulOutEncodedBytesLength = j;
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Init()
+{
+    #if ( LIBRARY_LOG_LEVEL >= LOG_INFO )
+        az_log_set_message_callback( prvAzureIoTLogListener );
+    #endif
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+void AzureIoT_Deinit()
+{
+    AZLogDebug( ( "AzureIoT_Deinit called" ) );
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_MessagePropertiesInit( AzureIoTMessageProperties_t * pxMessageProperties,
+                                                 uint8_t * pucBuffer,
+                                                 uint32_t ulAlreadyWrittenLength,
+                                                 uint32_t ulBufferLength )
+{
+    az_span xPropertyBufferSpan = az_span_create( pucBuffer, ( int32_t ) ulBufferLength );
+    az_result xResult;
+
+    if( pxMessageProperties == NULL )
+    {
+        AZLogError( ( "AzureIoT_MessagePropertiesInit failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
```

Comment:
Can we try to have one return from function.  Atleast in simple functions

Reply:
I understand the sentiment. Only problem I have is that gets to be wishy-washy with PR's. There will be debates about what constitutes "simple enough" for one return. Ideally I think we should either accept multiple returns or just have one. From a future maintainability standpoint.